Rules D204 cause autofix error

Ruff 0.0.287 (latest changes from main branch)
```
ruff  *.py --select D204 --no-cache --fix
```

file content(at least simple cpython script shows that this is valid python file):
```
class ServiceSorter(ABC):
	def sort_services(self,services):0
class DefaultPrioritySorter(ServiceSorter):
	'Default service sorting approach that always give S3 the highest priority.';priorities={'s3':100,'lambda':1}
	def sort_services(A,services):return sorted(services,key=lambda k:A.priorities.get(k,0),reverse=True)
```

error
```
/home/rafal/test/tmp_folder/39521.py:4:2: D204 1 blank line required after class docstring
Found 1 error.

error: Autofix introduced a syntax error. Reverting all changes.

This indicates a bug in `ruff`. If you could open an issue at:

    https://github.com/astral-sh/ruff/issues/new?title=%5BAutofix%20error%5D

...quoting the contents of `/home/rafal/test/tmp_folder/39521.py`, the rule codes D204, along with the `pyproject.toml` settings and executed command, we'd be very appreciative!


```
    
[python_compressed.zip](https://github.com/astral-sh/ruff/files/12506516/python_compressed.zip)

--- a/crates/ruff/resources/test/fixtures/pydocstyle/D.py
+++ b/crates/ruff/resources/test/fixtures/pydocstyle/D.py
@@ -644,3 +644,17 @@ def same_line(): """This is a docstring on the same line"""
 def single_line_docstring_with_an_escaped_backslash():
     "\
     "
+
+class StatementOnSameLineAsDocstring:
+    "After this docstring there's another statement on the same line separated by a semicolon." ; priorities=1
+    def sort_services(self):
+        pass
+
+class StatementOnSameLineAsDocstring:
+    "After this docstring there's another statement on the same line separated by a semicolon."; priorities=1
+
+
+class CommentAfterDocstring:
+    "After this docstring there's a comment."  # priorities=1
+    def sort_services(self):
+        pass

--- a/crates/ruff/src/rules/pydocstyle/rules/blank_before_after_class.rs
+++ b/crates/ruff/src/rules/pydocstyle/rules/blank_before_after_class.rs
@@ -1,7 +1,7 @@
 use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit, Fix};
 use ruff_macros::{derive_message_formats, violation};
-use ruff_python_trivia::PythonWhitespace;
-use ruff_source_file::{UniversalNewlineIterator, UniversalNewlines};
+use ruff_python_trivia::{indentation_at_offset, PythonWhitespace};
+use ruff_source_file::{Line, UniversalNewlineIterator};
 use ruff_text_size::Ranged;
 use ruff_text_size::{TextLen, TextRange};
 
@@ -216,25 +216,61 @@ pub(crate) fn blank_before_after_class(checker: &mut Checker, docstring: &Docstr
     }
 
     if checker.enabled(Rule::OneBlankLineAfterClass) {
-        let after_range = TextRange::new(docstring.end(), class.end());
+        let class_after_docstring_range = TextRange::new(docstring.end(), class.end());
+        let class_after_docstring = checker.locator().slice(class_after_docstring_range);
+        let mut lines = UniversalNewlineIterator::with_offset(
+            class_after_docstring,
+            class_after_docstring_range.start(),
+        );
 
-        let after = checker.locator().slice(after_range);
-
-        let all_blank_after = after.universal_newlines().skip(1).all(|line| {
+        // If the class is empty except for comments, we don't need to insert a newline between
+        // docstring and no content
+        let all_blank_after = lines.clone().all(|line| {
             line.trim_whitespace().is_empty() || line.trim_whitespace_start().starts_with('#')
         });
         if all_blank_after {
             return;
         }
 
-        let mut lines = UniversalNewlineIterator::with_offset(after, after_range.start());
+        let first_line = lines.next();
+        let mut replacement_start = first_line.as_ref().map(Line::start).unwrap_or_default();
 
-        let first_line_start = lines.next().map(|l| l.start()).unwrap_or_default();
+        // Edge case: There is trailing end-of-line content after the docstring, either a statement
+        // separated by a semicolon or a comment.
+        if let Some(first_line) = &first_line {
+            let trailing = first_line.as_str().trim_whitespace_start();
+            if let Some(next_statement) = trailing.strip_prefix(';') {
+                let indentation = indentation_at_offset(docstring.start(), checker.locator())
+                    .expect("Own line docstring must have indentation");
+                let mut diagnostic = Diagnostic::new(OneBlankLineAfterClass, docstring.range());
+                if checker.patch(diagnostic.kind.rule()) {
+                    let line_ending = checker.stylist().line_ending().as_str();
+                    // We have to trim the whitespace twice, once before the semicolon above and
+                    // once after the semicolon here, or we get invalid indents:
+                    // ```rust
+                    // class Priority:
+                    //     """Has priorities"""   ;   priorities=1
+                    // ```
+                    let next_statement = next_statement.trim_whitespace_start();
+                    diagnostic.set_fix(Fix::automatic(Edit::replacement(
+                        line_ending.to_string() + line_ending + indentation + next_statement,
+                        replacement_start,
+                        first_line.end(),
+                    )));
+                }
+                checker.diagnostics.push(diagnostic);
+                return;
+            } else if trailing.starts_with('#') {
+                // Keep the end-of-line comment, start counting empty lines after it
+                replacement_start = first_line.end();
+            }
+        }
+
         let mut blank_lines_after = 0usize;
-        let mut blank_lines_end = docstring.end();
+        let mut blank_lines_end = first_line.as_ref().map_or(docstring.end(), Line::end);
 
         for line in lines {
-            if line.trim().is_empty() {
+            if line.trim_whitespace().is_empty() {
                 blank_lines_end = line.end();
                 blank_lines_after += 1;
             } else {
@@ -248,7 +284,7 @@ pub(crate) fn blank_before_after_class(checker: &mut Checker, docstring: &Docstr
                 // Insert a blank line before the class (replacing any existing lines).
                 diagnostic.set_fix(Fix::automatic(Edit::replacement(
                     checker.stylist().line_ending().to_string(),
-                    first_line_start,
+                    replacement_start,
                     blank_lines_end,
                 )));
             }

--- a/crates/ruff_source_file/src/newlines.rs
+++ b/crates/ruff_source_file/src/newlines.rs
@@ -32,6 +32,7 @@ impl UniversalNewlines for str {
 /// assert_eq!(lines.next_back(), Some(Line::new("\r\n", TextSize::from(8))));
 /// assert_eq!(lines.next(), None);
 /// ```
+#[derive(Clone)]
 pub struct UniversalNewlineIterator<'a> {
     text: &'a str,
     offset: TextSize,

